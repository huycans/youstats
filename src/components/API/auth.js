const verifyToken = async (googleToken) => {
  let response = await fetch(
    "https://oauth2.googleapis.com/tokeninfo?id_token=" + googleToken
  );
  let responseJson = await response.json();
  if (responseJson.error) return false;
  return true;
};

export { verifyToken };
