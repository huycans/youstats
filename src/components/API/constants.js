export const googleURL = {
  googleOAuth: "https://oauth2.googleapis.com/tokeninfo",
  youtubeSearch: "https://www.googleapis.com/youtube/v3/search",
  youtubeChannelSearch: "https://youtube.googleapis.com/youtube/v3/channels",
  youtubePlaylistItemSearch:
    "https://youtube.googleapis.com/youtube/v3/playlistItems"
};
