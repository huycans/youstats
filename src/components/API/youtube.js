const fetchYoutubeChannelsByKeyword = async (keyword) => {
  const MAX_RESULT = 6;
  const params = `?part=snippet&type=channel&maxResults=${MAX_RESULT}&q=${keyword}&key=${process.env.REACT_APP_API_KEY}`;
  const response = await fetch(
    "https://www.googleapis.com/youtube/v3/search" + params,
    {
      method: "GET",
      headers: {
        Accept: "application/json"
      }
    }
  );
  return await response.json();
};

export const youtubeAPI = { fetchYoutubeChannelsByKeyword };
