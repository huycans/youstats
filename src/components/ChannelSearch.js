import React, { useState, useEffect } from "react";
import { useSelector, useDispatch } from "react-redux";
import {
  fetchYoutubeChannelsByKeyword,
  fetchYoutubeChannelInfo
} from "./slices/youtube";
import { useDebounce } from "./utils";

export default function ChannelSearch() {
  const debouncedTime = 500;
  const showChannelResults = useSelector(
    (state) => state.youtube.showChannelResults
  );
  const channelResults = useSelector((state) => state.youtube.channelResults);
  let [searchTerm, setSearchTerm] = useState("");

  let debouncedSearchTerm = useDebounce(searchTerm, debouncedTime);

  const dispatch = useDispatch();

  //handle calling API when debouncedSearchTerm changes
  useEffect(() => {
    if (debouncedSearchTerm.length > 1) {
      dispatch(fetchYoutubeChannelsByKeyword(debouncedSearchTerm));
    }
  }, [debouncedSearchTerm, dispatch]);

  //handle user typing
  const handleChange = (event) => {
    setSearchTerm(event.target.value);
  };

  const handleChannelClick = (e, channelId) => {
    dispatch(fetchYoutubeChannelInfo(channelId));
  };

  return (
    <>
      <div className="channel-search">
        <label htmlFor="channel-search-input">
          {" "}
          Search for a channel by name: &nbsp;&nbsp;
          <input
            data-testid="channel-search-input"
            id="channel-search-input"
            type="text"
            onChange={handleChange}
          />
        </label>
      </div>
      <ul className="channel-results-lists" data-testid="channel-result-lists">
        {showChannelResults
          ? channelResults.map((result, index) => {
              return (
                <li
                  data-testid="channel-result-items"
                  className="container channel-results-item"
                  key={result.id.channelId}
                  onClick={(e) => handleChannelClick(e, result.id.channelId)}
                >
                  <div className="row">
                    <div className="col-2">
                      <img
                        className="channel-results-logo"
                        alt={result.snippet.thumbnails.title}
                        src={result.snippet.thumbnails.default.url}
                        data-testid={"channel-result-items-logo"}
                      />
                    </div>
                    <div className="col-10">
                      <div
                        className="row channel-title"
                        data-testid={"channel-result-items-title"}
                      >
                        {result.snippet.channelTitle}
                      </div>
                      <div className="row channel-">
                        {result.snippet.description.slice(0, 80) + "..."}
                      </div>
                    </div>
                  </div>
                </li>
              );
            })
          : null}
      </ul>
    </>
  );
}
