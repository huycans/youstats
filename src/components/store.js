import { configureStore } from "@reduxjs/toolkit";
import authReducer from "./slices/auth";
import displayReducer from "./slices/display";
import youtubeReducer from "./slices/youtube";

export default configureStore({
  reducer: {
    auth: authReducer,
    youtube: youtubeReducer,
    display: displayReducer
  }
});
