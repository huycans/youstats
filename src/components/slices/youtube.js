import { createSlice, createAsyncThunk } from "@reduxjs/toolkit";
import * as _ from "lodash";
import { youtubeAPI } from "../API/youtube";

const initialState = {
  channelResults: [],
  showChannelResults: false,
  currentChannel: {
    isEmpty: true,
    title: "",
    description: "",
    thumbnail: "",
    onYoutubeSince: "",
    country: "",
    uploads: "",
    viewCount: "",
    subscriberCount: "",
    videoCount: "",
    topics: [],
    kidsFriendly: false,
    bannerURL: ""
  },
  latestVideoId: ""
};

//thunk action
export const fetchYoutubeChannelsByKeyword = createAsyncThunk(
  "youtube/fetchYoutubeChannelsByKeyword",
  async (keyword, thunkAPI) => {
    try {
      return await youtubeAPI.fetchYoutubeChannelsByKeyword(keyword);
    } catch (err) {
      return thunkAPI.rejectWithValue(err.response.data);
    }
  }
);

export const fetchYoutubeChannelInfo = createAsyncThunk(
  "youtube/fetchYoutubeChannelInfo",
  async (channelId, thunkAPI) => {
    try {
      return await youtubeAPI.fetchYoutubeChannelInfo(channelId);
    } catch (err) {
      return thunkAPI.rejectWithValue(err.response.data);
    }
  }
);

export const fetchYoutubePlaylistById = createAsyncThunk(
  "youtube/fetchYoutubePlaylistById",
  async (playlistId, thunkAPI) => {
    try {
      return await youtubeAPI.fetchYoutubePlaylistById(playlistId);
    } catch (err) {
      return thunkAPI.rejectWithValue(err.response.data);
    }
  }
);
//END thunk action

export const youtubeSlice = createSlice({
  name: "youtube",
  initialState: initialState,
  reducers: {},
  extraReducers: (builder) => {
    builder
      .addCase(fetchYoutubeChannelsByKeyword.fulfilled, (state, action) => {
        state.channelResults = action.payload.items;
        state.showChannelResults = true;
      })
      .addCase(fetchYoutubeChannelInfo.fulfilled, (state, action) => {
        const resultObj = action.payload.items[0];

        const getPropertyFromObjectCurried = function (obj, defaultValue) {
          //this util function check if an object has the given path, if yes return the value of that path
          //using curried function to define an obj and a default value first
          //then return another function that will receive a path
          //finally return the value of that path inside the obj
          return function (path) {
            return _.get(obj, path, defaultValue);
          };
        };
        const getPropertyAtPath = getPropertyFromObjectCurried(resultObj, "");
        state.showChannelResults = false;
        state.currentChannel = {
          isEmpty: false,
          channelId: getPropertyAtPath("id.channelId"),
          title: getPropertyAtPath("snippet.title"),
          description: getPropertyAtPath("snippet.description"),
          thumbnailDefault: getPropertyAtPath("snippet.thumbnails.default.url"),
          thumbnailBig: getPropertyAtPath("snippet.thumbnails.high.url"),
          onYoutubeSince: getPropertyAtPath("snippet.publishedAt"),
          country: getPropertyAtPath("snippet.country"),
          uploads: getPropertyAtPath("contentDetails.relatedPlaylists.uploads"),
          viewCount: getPropertyAtPath("statistics.viewCount"),
          subscriberCount: getPropertyAtPath("statistics.subscriberCount"),
          videoCount: getPropertyAtPath("statistics.videoCount"),
          topics: getPropertyFromObjectCurried(
            resultObj,
            []
          )("topicDetails.topicIds"),
          kidsFriendly: getPropertyAtPath("status.madeForKids"),
          bannerURL: getPropertyAtPath(
            "brandingSettings.image.bannerExternalUrl"
          )
        };
      })
      .addCase(fetchYoutubePlaylistById.fulfilled, (state, action) => {
        state.latestVideoId =
          action.payload.items[0].snippet.resourceId.videoId;
      });
  }
});

// export const {  } = youtubeSlice.actions;
export default youtubeSlice.reducer;
