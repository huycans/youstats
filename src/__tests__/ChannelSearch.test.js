import React from "react";
import renderer from "react-test-renderer";
import ChannelSearch from "../components/ChannelSearch";
import { Provider } from "react-redux";
import store from "../components/store";
import { render, cleanup, fireEvent } from "@testing-library/react";
import { youtubeAPI } from "../components/API/youtube";

jest.mock("../components/API/youtube");

jest.mock("../components/utils.js", () => {
  return {
    useDebounce: (text) => {
      return text;
    }
  };
});

describe("<NavigationBar />", () => {
  let WrappedChannelSearch;
  beforeEach(() => {
    WrappedChannelSearch = () => (
      <Provider store={store}>
        <ChannelSearch />
      </Provider>
    );
  });

  afterEach(() => {
    cleanup();
  });

  it("renders correctly", () => {
    const tree = renderer.create(<WrappedChannelSearch />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("renders the input element", () => {
    const { queryByTestId } = render(<WrappedChannelSearch />);

    expect(queryByTestId("channel-search-input")).toBeTruthy();
  });

  it("renders two dummy items after user type on the input element", async () => {
    const { queryByTestId, findAllByTestId } = render(<WrappedChannelSearch />);
    const input = queryByTestId("channel-search-input");
    fireEvent.change(input, "random text");
    // await new Promise(setImmediate);
    let searchResults = await findAllByTestId("channel-result-items");
    expect(searchResults.length).toBe(2);
  });
});
