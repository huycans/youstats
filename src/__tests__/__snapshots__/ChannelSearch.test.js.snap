// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NavigationBar /> renders correctly 1`] = `
<div
  className="col-12 channel-search-container col-lg-6"
>
  <div
    className="row channel-search"
  >
    <h3>
      Search for a youtube channel by name
    </h3>
    <input
      className="channel-search-input"
      data-testid="channel-search-input"
      id="channel-search-input"
      onChange={[Function]}
      type="text"
    />
  </div>
</div>
`;
